Add file descriptor set service documentation (#277)
bufbuild/docs.buf.build: docs/bsr/api-access.md

---
id: api-access
title: API Access
---

Buf Schema Registries (BSRs for short) provide a web UI at their domain. For example, the web

> UI for the public BSR is [https://buf.build](https://buf.build/). (This is also Buf's main

Reviewer: delete `https://` from the displayed text. Also, tone down the parentheses :-)

Reviewer 2: Done

Connect, [gRPC](https://grpc.io/), and [gRPC-Web](https://github.com/grpc/grpc-web#readme).

Since the APIs are defined in Protobuf, you can use `buf generate` to generate an API client in
a variety of languages. Connect support is still growing, so you can use gRPC clients if you
use a language not yet supported. You can also use [Remote Packages](/bsr/remote-packages/overview)
to download a client from the BSR.

The [Connect protocol](https://connect.build/docs/protocol) also makes it easy to consume
non-streaming (aka “unary”) endpoints from clients that, for whatever reason, cannot use a
generated Connect or gRPC client. For example, you can even use `curl` or `wget` with unary
endpoints. Connect supports JSON encoding of request and response messages out-of-the-box.
Since the API is defined in Protobuf, refer to the
[official JSON mapping documentation](https://protobuf.dev/programming-guides/proto3/#json)
for how a message is represented in JSON.

All APIs are currently exposed at the root URI (currently required for gRPC compatibility).
For example, the URL to use for an RPC to the public BSR looks like so:

> > https://api<area/>.buf.build/_&lt;fully.qualified.ServiceName>/&lt;MethodName>_

Reviewer 2: This is gross, but was needed so that
1. It wasn't turned into a (broken) link since it looks like a URL. The `<area/>` in the middle is what does that. When it _does_ think the whole thing is a link, it also treats the underscores as part of the URL path (instead of as indicator for italics).
2. So that text inside angle brackets wasn't misinterpreted as an HTML tag, I had to change the opening bracket to an HTML entity.

Ugly in source, but renders nicely (IMO)

## Authentication

Most Buf Schema Registry APIs will require the client to be authenticated. Before you begin,
[get an API token](authentication#create-an-api-token) from the BSR settings page. Once you've
got a token, put it in the `Authorization` request header with a schema of `Bearer`, and that
will authenticate your request with a BSR. So the final HTTP request headers should look like
so:

> > Authorization: Bearer _&lt;token&gt;_

Author: ```suggestion
> Authorization: Bearer _&lt;token&gt;_
```
is this a notion->github format issue?

Reviewer 2: Yes. Well, notion -> docusaurus (but also most other Markdown renderers don't like it either, because they think it's an HTML tag). It turns out, you need to put stuff in backticks `` to avoid HTML interpretation or use HTML entities.

## Examples

```bash
curl \
   https://api.buf.build/buf.reflect.v1beta1.FileDescriptorSetService/GetFileDescriptorSet \
   -H "Authorization: Bearer ${BUF_TOKEN}" \
   -H "Content-Type: application/json" \
   -X POST -d '{"module": "buf.build/bufbuild/eliza"}'
```

cURL is a great way for us to understand how to access the APIs, but it's unlikely to be

> useful in code. For that, see an example using another feature of the BSR called

Author: ```suggestion
useful in production. For that, see an example using another feature of the BSR called
```
?

Reviewer 2: Hmm. I thought I had edited that to `application code`.

> useful in code. For that, see an example using another feature of the BSR called

Reviewer 2: ```suggestion
useful in code. For that, let's look at an example using another feature of the BSR called
```
